src/Build/Globbing/MSBuildGlob.cs

```diff
@@ -198,11 +203,28 @@ public static MSBuildGlob Parse(string globRoot, string fileSpec)
                         return (normalizedFixedPart, wildcardDirPart, filePart);
                     });
 
-                // compile the regex since it's expected to be used multiple times
-                var regex = isLegalFileSpec
-                    ? new Regex(matchFileExpression, FileMatcher.DefaultRegexOptions | RegexOptions.Compiled)
-                    : null;
+                Regex regex = null;
+                if (isLegalFileSpec)
+                {
+                    lock (s_regexCache)
+                    {
+                        s_regexCache.TryGetValue(matchFileExpression, out regex);
+                    }
 
+                    if (regex == null)
+                    {
+                        // compile the regex since it's expected to be used multiple times
+                        Regex newRegex = new Regex(matchFileExpression, FileMatcher.DefaultRegexOptions | RegexOptions.Compiled);
```

Comment:
How much pressure do you expect to be on the lock? If not too much, you could move the Regex creation into the lock and avoid creating it when it turns out to not be necessary (an edge case, admittedly).

Reply:
This is a public API and the usage pattern can be anything, really, so I wanted to err on the side of caution. Creating `Regex` when it's not really necessary is something that's been happening until now at a large scale so nothing should break if there's still a rare occurrence. On the other hand, holding the lock while compiling the regex could hypothetically regress concurrent scenarios. Btw, similar reason why I'm caching `Regex` and not the entire `MSBuildGlob` even though the latter is also immutable. Saving the small object just doesn't seem worth the potential break. Let me know if I'm overthinking it .. when I see the `public` modifier I tend to expect the worst.

Reply:
This is probably overoptimizing, but rather than locking the DS, you could lock parts of the DS. Looking at our [ConcurrentDictionary](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1900), I see that each bucket has an associated lock, so that dramatically reduces pressure from parallelism. I'm curious why they didn't just make one lock per bucket, but they probably wanted to minimize memory pressure, and this is a reasonable compromise.

In any case, if you were to make a ConcurrentWeakValueDictionary based on ConcurrentDictionary and use that instead, you don't have to lock the whole ds, and you probably wouldn't have to worry about the time it takes to compile a regex or maybe even the whole glob. What do you think?

Reply:
What you're suggesting makes sense and would indeed help parallelize the code. Looking at the usage pattern of the API in Visual Studio (most calls are made on the same thread and the total number of `Regex` objects is in the order of 100s, even for very large solutions), I don't believe that there is a need for it, though, and making a ConcurrentWeakValueDictionary would not be justified.